### PR TITLE
Consider customized top-level type names

### LINKED ISSUE #135: Customized top-level type names are not recognized
The `CustomQueryType` of the following schema is not enhanced, instead a new `Query` type is generated:

```graphql
type Person { name: String }
type Location { name: String }

type CustomQueryType {
  person(name:String) : Person @cypher(statement:"MATCH (p:Person { name:{name} }) RETURN p")
}

schema {
  query: CustomQueryType
}
```

Current target schema:

```graphql
schema {
  query: Query
  mutation: Mutation
}

type CustomQueryType {
  person(name: String): Person @cypher(statement : "MATCH (p:Person { name:{name} }) RETURN p")
}

type Location {
  name: String
}

type Mutation {
  createLocation(name: String): Location!
  createPerson(name: String): Person!
}

type Person {
  name: String
}

type Query {
  location(filter: _LocationFilter, first: Int, name: String, offset: Int, orderBy: [_LocationOrdering!]): [Location!]!
  person(filter: _PersonFilter, first: Int, name: String, offset: Int, orderBy: [_PersonOrdering!]): [Person!]!
}
```

expected:

```graphql
schema {
  query: CustomQueryType
  mutation: Mutation
}

type CustomQueryType {
  location(filter: _LocationFilter, first: Int, name: String, offset: Int, orderBy: [_LocationOrdering!]): [Location!]!
  person(name: String): Person @cypher(statement : "MATCH (p:Person { name:{name} }) RETURN p")
}

type Location {
  name: String
}

type Mutation {
  createLocation(name: String): Location!
  createPerson(name: String): Person!
}

type Person {
  name: String
}
```

### PR DESCRIPTION
**Pull request by Andy2003**
resolves #135

### CODE REVIEW

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 1]
```diff
@@ -74,11 +70,34 @@ object SchemaBuilder {
...
         val handler = getHandler(config)
 
-        var targetSchema = augmentSchema(sourceSchema, handler, config)
+        var targetSchema = augmentSchema(sourceSchema, handler)
         targetSchema = addDataFetcher(targetSchema, dataFetchingInterceptor, handler)
         return targetSchema
     }
 
+    private fun ensureRootQueryTypeExists(enhancedRegistry: TypeDefinitionRegistry) {
+        val queryCreator: (SchemaDefinition.Builder) -> Unit = { t: SchemaDefinition.Builder ->
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt:79]: can't this just be a nested fun?

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 2]
```diff
@@ -74,11 +70,34 @@ object SchemaBuilder {
...
+                .typeName(TypeName("Query"))
+                .build())
+        }
+        val queryDefinition = ObjectTypeDefinition.newObjectTypeDefinition().name("Query").build()
+
+        val schemaDefinition = enhancedRegistry.schemaDefinition().orElse(null)
+        if (schemaDefinition == null) {
+            enhancedRegistry.add(queryDefinition)
+            val newSchemaDefinition = SchemaDefinition.newSchemaDefinition()
+            queryCreator(newSchemaDefinition)
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt:92]: why not sd.transform here?

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 3]
```diff
@@ -99,18 +118,20 @@ object SchemaBuilder {
...
-    private fun augmentSchema(sourceSchema: GraphQLSchema, handler: List<AugmentationHandler>, config: SchemaConfig): GraphQLSchema {
+    private fun augmentSchema(sourceSchema: GraphQLSchema, handler: List<AugmentationHandler>): GraphQLSchema {
         val types = sourceSchema.typeMap.toMutableMap()
-        val env = BuildingEnv(types)
-
+        val env = BuildingEnv(types, sourceSchema)
+        val queryTypeName = sourceSchema.queryTypeName()
+        val mutationTypeName = sourceSchema.mutationTypeName()
+        val subscriptionTypeName = sourceSchema.subscriptionTypeName()
         types.values
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt:127]: in general it would be cool if we could have more transformations than in-place updates

**Review by jexp — APPROVED** [review]

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt (hunk 3 above)
Comment (Andy2003) [reply src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt:127]: I will pay attention on future refactorings

Comment (jexp) [reply src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt:128]: Thanks!

**Review by jexp — APPROVED** [review]

**Pull request merged**